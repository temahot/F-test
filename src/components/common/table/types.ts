import { type UsePaginationReturn } from '~/hooks/usePagination';

export interface TableProps {
  data: Array<Record<string, any>> | [];
  headers: string[];
  pagination?: UsePaginationReturn['pagination'];
  itemsAmount?: number;
  isLoading: boolean;
}
