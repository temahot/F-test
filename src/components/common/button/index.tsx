import React from 'react';

import { Button as MuiButton, type ButtonProps } from '@mui/material';

interface ExpandedButtonProps extends ButtonProps {
  disabled?: boolean;
  isLoading?: boolean;
}
export const Button = ({
  isLoading = false,
  disabled,
  ...rest
}: ExpandedButtonProps) => (
  <MuiButton {...rest} disabled={disabled ?? isLoading} />
);
