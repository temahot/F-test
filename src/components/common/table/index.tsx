import React, { ReactElement } from 'react';

import {
  Paper,
  Table as MuiTable,
  TableContainer,
  TablePagination,
  TableRow,
} from '@mui/material';

import { TableBody } from '~/components/common/table/body';
import { TableHead } from '~/components/common/table/head';
import { type TableProps } from '~/components/common/table/types';

export const Table = ({
  data,
  headers,
  pagination,
  itemsAmount,
  isLoading,
}: TableProps): ReactElement => {
  return (
    <TableContainer component={Paper}>
      <MuiTable>
        <TableHead headers={headers} />
        <TableBody data={data} isLoading={isLoading} headers={headers}>
          {pagination && itemsAmount && data.length !== 0 ? (
            <TableRow>
              <TablePagination
                count={-1}
                page={pagination.page}
                onPageChange={(_, p) => {
                  pagination.setPage(p);
                }}
                nextIconButtonProps={{
                  disabled:
                    itemsAmount <=
                    pagination.page * pagination.rowsPerPage +
                      pagination.rowsPerPage,
                }}
                backIconButtonProps={{
                  disabled: pagination.page === 1,
                }}
                rowsPerPage={pagination.rowsPerPage}
                rowsPerPageOptions={[]}
              />
            </TableRow>
          ) : null}
        </TableBody>
      </MuiTable>
    </TableContainer>
  );
};
