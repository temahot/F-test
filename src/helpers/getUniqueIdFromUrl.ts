const getUniqueIdFromUrl = (url: string) => {
  const regex = /people\/(\d+)/;
  const match = url.match(regex);
  return match ? match[1] : url;
};

export default getUniqueIdFromUrl;
