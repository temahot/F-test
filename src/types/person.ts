export type Person = {
  birth_year: string;
  eye_color: string;
  gender: string;
  hair_color: string;
  height: string;
  homeworld: string;
  mass: string;
  name: string;
  skin_color: string;
  url: string;
};

export type PersonPaginatedResponse = {
  count: number;
  next: string | null;
  previous: string | null;
  results: Person[];
};
