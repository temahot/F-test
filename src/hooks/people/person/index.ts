import { useMemo } from 'react';

import { useRouter } from 'next/router';
import useSWR from 'swr';

import { notify } from '~/helpers/common/notify';
import useNetwork from '~/hooks/common/useNetwork';
import { ROUTES } from '~/routes';
import { type Person } from '~/types/person';

interface UsePersonReturn {
  personIsLoading: boolean;
  person: Person | undefined;
}

interface UsePersonProps {
  id: number | string;
}

export default function usePerson({ id }: UsePersonProps): UsePersonReturn {
  const { api } = useNetwork();
  const router = useRouter();
  const fetchPersonById = async () => {
    try {
      const response = await api.people.fetchPeopleById({
        id,
      });
      return response.data;
    } catch {
      notify('error', 'Person not found');
      await router.push(`/${ROUTES.PEOPLE}`);
      throw new Error('fetchPersonById failed');
    }
  };

  const { data, isLoading } = useSWR<Person>(
    `${ROUTES.PEOPLE}/${id}`,
    fetchPersonById,
    {
      revalidateOnFocus: false,
      revalidateOnMount: false,
    }
  );

  return useMemo(
    () => ({
      personIsLoading: isLoading,
      person: data,
    }),
    [data, isLoading]
  );
}
