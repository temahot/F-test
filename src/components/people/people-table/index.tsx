import React, { ReactElement, useCallback, useMemo, useState } from 'react';

import { styled } from '@mui/material';
import Typography from '@mui/material/Typography';
import { useRouter } from 'next/router';
import { useSWRConfig } from 'swr';

import { Table } from '~/components/common/table';
import { TextFilter } from '~/components/text-filter';
import getUniqueIdFromUrl from '~/helpers/getUniqueIdFromUrl';
import { FILTER_KEYS } from '~/hooks/usePagination';
import usePeople from '~/hooks/usePeople';
import { ROUTES } from '~/routes';
import { COLORS } from '~/styles/colors';
import { type Person } from '~/types/person';

const NameLink = styled(Typography)({
  cursor: 'pointer',
  color: COLORS.PRIMARY,
  textDecoration: 'underline',
});

export interface OuterFilters {
  search: string;
}

type TableHeaders = Pick<
  Person,
  'name' | 'skin_color' | 'height' | 'mass' | 'birth_year'
> & { id: string };

const headers: Record<keyof TableHeaders, string> = {
  id: 'Id',
  name: 'Name',
  skin_color: 'Color of skin',
  height: 'Height',
  mass: 'Mass',
  birth_year: 'Day of birth',
};

export const formattedHeaders = Object.values(headers);

export const PeopleTable = (): ReactElement => {
  const initialFilters = {
    search: '',
  };

  const [appliedFilters, setAppliedFilters] =
    useState<OuterFilters>(initialFilters);

  const { mutate } = useSWRConfig();
  const router = useRouter();

  const { people, peopleAreLoading, peopleAmount, pagination } = usePeople({
    outerFilters: appliedFilters,
  });

  const handleOpenPerson = useCallback(
    async (person: Person) => {
      const id = getUniqueIdFromUrl(person.url);
      void router.push({
        pathname: `${ROUTES.PEOPLE}/[id]`,
        query: { id },
      });
      await mutate(`${ROUTES.PEOPLE}/${id}`, person, {
        revalidate: false,
        optimisticData: true,
      });
    },
    [mutate, router],
  );

  const peopleTableUi = useMemo(
    () =>
      people.map((person) => ({
        id: getUniqueIdFromUrl(person.url),
        name: (
          <NameLink
            variant={'body1'}
            onClick={() => {
              void handleOpenPerson(person);
            }}
          >
            {person.name}
          </NameLink>
        ),
        skin_color: person.skin_color ?? 'empty',
        height: person.height ?? 'empty',
        mass: person.mass ?? 'empty',
        birth_year: person.birth_year ?? 'empty',
      })),
    [handleOpenPerson, people],
  );

  return (
    <>
      <TextFilter
        label={'Search by name'}
        onChange={({ target: { value } }) => {
          setAppliedFilters((prevState) => ({
            ...prevState,
            [FILTER_KEYS.SEARCH]: value,
          }));
        }}
      />

      <Table
        data={peopleTableUi}
        headers={formattedHeaders}
        itemsAmount={peopleAmount}
        pagination={pagination}
        isLoading={peopleAreLoading}
      />
    </>
  );
};
