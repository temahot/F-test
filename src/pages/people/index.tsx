import React from 'react';

import { PageTitle } from '~/components/page-title';
import { PeopleTable } from '~/components/people/people-table';

export default function People() {
  return (
    <>
      <PageTitle title={'People'} />
      <PeopleTable />
    </>
  );
}
