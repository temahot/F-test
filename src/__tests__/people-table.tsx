import React from 'react';
import renderer, { act } from 'react-test-renderer';

import { render, screen } from '@testing-library/react';
import { useRouter } from 'next/router';

import allPeopleResponse from '~/__mocks__/all-people-response';
import { Table } from '~/components/common/table';
import { formattedHeaders } from '~/components/people/people-table';
import getUniqueIdFromUrl from '~/helpers/getUniqueIdFromUrl';
import { type Person } from '~/types/person';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

let peopleResponse: Person[];

describe('Table component', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
    peopleResponse = allPeopleResponse.results.map((person) => ({
      ...person,
      id: getUniqueIdFromUrl(person.url),
    }));

    (useRouter as jest.Mock).mockReturnValue({ query: {} });
    await act(() => {
      render(
        <Table
          data={peopleResponse}
          headers={formattedHeaders}
          isLoading={false}
        />
      );
    });
  });

  it('Renders correctly', () => {
    const tree = renderer
      .create(
        <Table
          data={peopleResponse}
          headers={formattedHeaders}
          isLoading={false}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Mock data renders in columns', () => {
    const people = peopleResponse.map((person) =>
      screen.getByText(person.name)
    );

    people.forEach((item) => {
      expect(item).toBeTruthy();
    });
  });
});
