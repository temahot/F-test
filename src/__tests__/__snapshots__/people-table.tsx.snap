// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component Renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
>
  <table
    className="MuiTable-root css-rqglhn-MuiTable-root"
    role={null}
  >
    <thead
      className="MuiTableHead-root css-13rqw8u-MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-hover MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
          scope="col"
        >
          Id
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
          scope="col"
        >
          Name
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
          scope="col"
        >
          Color of skin
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
          scope="col"
        >
          Height
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
          scope="col"
        >
          Mass
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
          scope="col"
        >
          Day of birth
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-hover css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          Luke Skywalker
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          172
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          77
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          blond
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          fair
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          blue
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          19BBY
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          male
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          https://swapi.dev/api/planets/1/
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          https://swapi.dev/api/people/1/
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          1
        </td>
      </tr>
      <tr
        className="MuiTableRow-root MuiTableRow-hover css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          Darth Vader
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          202
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          136
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          none
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          white
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          yellow
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          41.9BBY
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          male
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          https://swapi.dev/api/planets/1/
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          https://swapi.dev/api/people/4/
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          4
        </td>
      </tr>
      <tr
        className="MuiTableRow-root MuiTableRow-hover css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          Leia Organa
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          150
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          49
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          brown
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          light
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          brown
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          19BBY
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          female
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          https://swapi.dev/api/planets/2/
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          https://swapi.dev/api/people/5/
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          5
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
